src/main/java/com/couchbase/lite/support/MultipartDocumentReader.java

```diff
@@ -37,26 +39,50 @@ public MultipartDocumentReader(HttpResponse response, Database database) {
     }
 
     public void parseJsonBuffer() {
+        byte[] json = jsonBuffer.toByteArray();
+
+        if(jsonCompressed){
+            json = Utils.decompressByGzip(json);
+            if(json == null){
+                throw new IllegalStateException("Received corrupt gzip-encoded JSON part");
+            }
+        }
+
         try {
-            document = Manager.getObjectMapper().readValue(jsonBuffer.toByteArray(), Map.class);
+            document = Manager.getObjectMapper().readValue(json, Map.class);
         } catch (IOException e) {
             throw new IllegalStateException("Failed to parse json buffer", e);
         }
+
         jsonBuffer = null;
     }
 
-    public void setContentType(String contentType) {
-        if (contentType.startsWith("multipart/")) {
+    public void setHeaders(Map<String, String> headers){
+        String contentType = headers.get("Content-Type");
+        if(contentType.startsWith("multipart/")){
+            // Multipart, so initialize the parser:
             multipartReader = new MultipartReader(contentType, this);
             attachmentsByName = new HashMap<String, BlobStoreWriter>();
```

Comment:
This line `attachmentsByName = new HashMap<String, BlobStoreWriter>();` seems to be duplicated